File: sdk/runanywhere-swift/Sources/RunAnywhere/Foundation/Errors/SDKException.swift
//
//  SDKException.swift
//  RunAnywhere
//
//  Wave 2: Canonical Swift error type. Wraps the generated proto `RASDKError`
//  (Sources/RunAnywhere/Generated/errors.pb.swift) so Swift `throws` works
//  while keeping the wire-canonical proto as the source of truth.
//
//  Replaces the legacy `SDKError` struct, `ErrorCode` String enum, and
//  `ErrorCategory` String enum. There is NO `typealias SDKError = SDKException`.
//  Every reference to the legacy types has been rewritten to use this one.
//
//  Usage:
//      throw SDKException.modelNotFound("whisper-base")
//      throw SDKException.stt(.notInitialized, "STT model not loaded")
//      do { ... } catch let ex as SDKException { print(ex.proto.message) }
//

import Foundation

// MARK: - SDKException

/// Canonical Swift-throwable error wrapping the generated proto `RASDKError`.
public struct SDKException: Error, LocalizedError, Sendable, CustomStringConvertible {

    /// The canonical proto-encoded error this exception wraps.
    public let proto: RASDKError

    /// Optional underlying Swift error (not part of the wire proto).
    public let underlying: (any Error)?

    /// Stack trace captured at construction time (debug aid; not on the wire).
    public let stackTrace: [String]

    public init(proto: RASDKError, underlying: (any Error)? = nil) {
        self.proto = proto
        self.underlying = underlying
        self.stackTrace = Thread.callStackSymbols
    }

    public init(
        code: RAErrorCode,
        message: String,
        category: RAErrorCategory = .component,
        underlying: (any Error)? = nil
    ) {
        var p = RASDKError()
        p.code = code
        p.message = message
        p.category = category
        // Round-trip C ABI code: positive proto code ↔ negative rac_result_t
        let raw = code.rawValue
        if raw > 0 && raw <= 899 {
            p.cAbiCode = -Int32(raw)
        }
        if let u = underlying {
            p.nestedMessage = String(describing: u)
        }
        self.proto = p
        self.underlying = underlying
        self.stackTrace = Thread.callStackSymbols
    }

    // MARK: Convenience accessors

    public var code: RAErrorCode { proto.code }
    public var category: RAErrorCategory { proto.category }
    public var message: String { proto.message }

    // MARK: LocalizedError

    public var errorDescription: String? { proto.message }

    public var failureReason: String? {
        "[\(proto.category)] \(proto.code)"
    }

    public var recoverySuggestion: String? {
        switch proto.code {
        case .notInitialized:
            return "Initialize the component before using it."
        case .modelNotFound:
            return "Ensure the model is downloaded and the path is correct."
        case .networkUnavailable:
            return "Check your internet connection and try again."
        case .insufficientStorage:
            return "Free up storage space and try again."
        case .insufficientMemory:
            return "Close other applications to free up memory."
        case .microphonePermissionDenied:
            return "Grant microphone permission in Settings."
        case .timeout:
            return "Try again or check your connection."
        case .invalidApiKey:
            return "Verify your API key is correct."
        case .cancelled:
            return nil
        default:
            return nil
        }
    }

    // MARK: CustomStringConvertible

    public var description: String {
        var result = "SDKException[\(proto.category).\(proto.code)]: \(proto.message)"
        if let u = underlying {
            result += "\n  Caused by: \(u)"
        }
        return result
    }

    /// Telemetry-only properties (lightweight, safe to ship).
    public var telemetryProperties: [String: String] {
        [
            "error_code": "\(proto.code)",
            "error_category": "\(proto.category)",
            "error_message": proto.message
        ]
    }
}

// MARK: - Equatable / Hashable

extension SDKException: Equatable {
    public static func == (lhs: SDKException, rhs: SDKException) -> Bool {
        lhs.proto.code == rhs.proto.code &&
        lhs.proto.category == rhs.proto.category &&
        lhs.proto.message == rhs.proto.message
    }
}

extension SDKException: Hashable {
    public func hash(into hasher: inout Hasher) {
        hasher.combine(proto.code.rawValue)
        hasher.combine(proto.category.rawValue)
        hasher.combine(proto.message)
    }
}

// MARK: - Generic factory

extension SDKException {
    /// Generic factory; auto-logs unexpected errors.
    public static func make(
        code: RAErrorCode,
        message: String,
        category: RAErrorCategory = .component,
        underlying: (any Error)? = nil,
        shouldLog: Bool = true
    ) -> SDKException {
        let ex = SDKException(code: code, message: message, category: category, underlying: underlying)
        if shouldLog && !code.isExpected {
            ex.log()
        }
        return ex
    }
}

// MARK: - Category-specific factories
//
// These mirror the legacy SDKError surface so callers like
// `SDKException.stt(.notInitialized, "msg")` continue to read naturally.
// Internally each one routes to `make(code:message:category:...)` with the
// proto's 9-bucket category enum.

extension SDKException {

    /// Generic / general SDK error.
    public static func general(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .internal, underlying: underlying)
    }

    /// Speech-to-Text component error.
    public static func stt(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// Text-to-Speech component error.
    public static func tts(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// LLM component error.
    public static func llm(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// VAD component error.
    public static func vad(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// VLM component error.
    public static func vlm(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// Speaker diarization component error.
    public static func speakerDiarization(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// Wake-word component error.
    public static func wakeWord(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// Voice agent component error.
    public static func voiceAgent(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// RAG component error.
    public static func rag(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// Diffusion component error.
    public static func diffusion(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .component, underlying: underlying)
    }

    /// Download / model-management error.
    public static func download(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .network, underlying: underlying)
    }

    /// File management error.
    public static func fileManagement(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .io, underlying: underlying)
    }

    /// Network error.
    public static func network(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .network, underlying: underlying)
    }

    /// Authentication error.
    public static func authentication(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .auth, underlying: underlying)
    }

    /// Security error.
    public static func security(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .auth, underlying: underlying)
    }

    /// Runtime error.
    public static func runtime(
        _ code: RAErrorCode,
        _ message: String,
        underlying: (any Error)? = nil
    ) -> SDKException {
        make(code: code, message: message, category: .internal, underlying: underlying)
    }
}

// MARK: - Common shortcuts

extension SDKException {
    /// Common shortcut: model not found.
    public static func modelNotFound(_ id: String) -> SDKException {
        make(code: .modelNotFound, message: "Model not found: \(id)", category: .model)
    }

    /// Common shortcut: not initialized.
    public static func notInitialized(_ what: String) -> SDKException {
        make(code: .notInitialized, message: "\(what) is not initialized", category: .component)
    }

    /// Common shortcut: invalid configuration.
    public static func invalidConfiguration(_ message: String) -> SDKException {
        make(code: .invalidConfiguration, message: message, category: .configuration)
    }

    /// Common shortcut: validation failed.
    public static func validationFailed(_ message: String) -> SDKException {
        make(code: .validationFailed, message: message, category: .validation)
    }

    /// Common shortcut: cancelled.
    public static func cancelled(_ message: String = "Operation cancelled") -> SDKException {
        make(code: .cancelled, message: message, category: .internal, shouldLog: false)
    }

    /// Common shortcut: not implemented.
    public static func notImplemented(_ message: String) -> SDKException {
        make(code: .notImplemented, message: message, category: .internal)
    }

    /// Common shortcut: timeout.
    public static func timeout(_ message: String) -> SDKException {
        make(code: .timeout, message: message, category: .network)
    }

    /// Common shortcut: network error.
    public static func networkError(_ message: String) -> SDKException {
        make(code: .networkError, message: message, category: .network)
    }
}

// MARK: - Conversion from arbitrary Error

extension SDKException {
    /// Convert any `Error` into an `SDKException`. If already one, returns it
    /// unchanged. Otherwise wraps in an unknown / general error.
    public static func from(_ error: any Error, category: RAErrorCategory = .internal) -> SDKException {
        if let ex = error as? SDKException { return ex }

        let nsError = error as NSError
        if nsError.domain == NSURLErrorDomain {
            return fromURLError(nsError, category: category)
        }
        return make(
            code: .unknown,
            message: error.localizedDescription,
            category: category,
            underlying: error
        )
    }

    public static func from(_ error: (any Error)?, category: RAErrorCategory = .internal) -> SDKException {
        guard let error = error else {
            return make(code: .unknown, message: "Unknown error", category: category)
        }
        return from(error, category: category)
    }

    private static func fromURLError(_ nsError: NSError, category: RAErrorCategory) -> SDKException {
        let code: RAErrorCode
        switch nsError.code {
        case NSURLErrorNotConnectedToInternet, NSURLErrorNetworkConnectionLost:
            code = .networkUnavailable
        case NSURLErrorTimedOut:
            code = .timeout
        case NSURLErrorCancelled:
            code = .cancelled
        case NSURLErrorCannotFindHost, NSURLErrorCannotConnectToHost:
            code = .networkError
        default:
            code = .networkError
        }
        return make(
            code: code,
            message: nsError.localizedDescription,
            category: category,
            underlying: nsError
        )
    }
}

// MARK: - ONNX Runtime error mapping

extension SDKException {
    /// Map an ONNX Runtime C error code into an SDKException.
    public static func fromONNXCode(_ code: Int32) -> SDKException {
        switch code {
        case 0:
            return runtime(.unknown, "Unexpected success code passed to error handler")
        case -1:
            return runtime(.initializationFailed, "ONNX Runtime initialization failed")
        case -2:
            return runtime(.modelLoadFailed, "Failed to load ONNX model")
        case -3:
            return runtime(.generationFailed, "ONNX inference failed")
        case -4:
            return runtime(.invalidState, "Invalid ONNX handle")
        case -5:
            return runtime(.invalidInput, "Invalid ONNX parameters")
        case -6:
            return runtime(.insufficientMemory, "ONNX Runtime out of memory")
        case -7:
            return runtime(.notImplemented, "ONNX feature not implemented")
        case -8:
            return runtime(.cancelled, "ONNX operation cancelled")
        case -9:
            return runtime(.timeout, "ONNX operation timed out")
        case -10:
            return runtime(.storageError, "ONNX IO error")
        default:
            return runtime(.unknown, "ONNX error code: \(code)")
        }
    }
}

// MARK: - RAErrorCode classification helper

extension RAErrorCode {
    /// Whether this error is expected/routine and shouldn't be logged as error.
    public var isExpected: Bool {
        switch self {
        case .cancelled, .streamCancelled:
            return true
        default:
            return false
        }
    }
}

// MARK: - Logging hook

extension SDKException {
    /// Log this exception to all configured destinations.
    public func log(file: String = #file, line: Int = #line, function: String = #function) {
        let level: LogLevel = (proto.code == .cancelled) ? .info : .error
        let fileName = (file as NSString).lastPathComponent

        var metadata: [String: Any] = [ // swiftlint:disable:this prefer_concrete_types avoid_any_type
            "error_code": "\(proto.code)",
            "error_category": "\(proto.category)",
            "source_file": fileName,
            "source_line": line,
            "source_function": function
        ]

        if let underlying = underlying {
            metadata["underlying_error"] = String(describing: underlying)
        }
        if let reason = failureReason {
            metadata["failure_reason"] = reason
        }

        // Top SDK frames only (cheap and useful)
        let sdkFrames = stackTrace.filter { $0.contains("RunAnywhere") }.prefix(5)
        if !sdkFrames.isEmpty {
            metadata["stack_trace"] = sdkFrames.joined(separator: "\n")
        }

        Logging.shared.log(
            level: level,
            category: "\(proto.category)",
            message: proto.message,
            metadata: metadata
        )
    }
}

// MARK: - C ABI helpers

extension SDKException {
    /// Map a `rac_result_t` code to an SDKException, or nil on success.
    public var rawCABICode: Int32 {
        proto.cAbiCode
    }
}

File: sdk/runanywhere-web-next/packages/core/src/Foundation/SDKException.ts
import {
  errorCategoryToJSON,
  errorCodeToJSON,
  ErrorCategory as ProtoErrorCategory,
  ErrorCode as ProtoErrorCode,
  ErrorSeverity as ProtoErrorSeverity,
  type SDKError as ProtoSDKError,
} from '@runanywhere/proto-ts/errors';
import { SDKLogger } from './SDKLogger';

function categoryForCode(code: number): ProtoErrorCategory {
  if (code === 0) return ProtoErrorCategory.ERROR_CATEGORY_UNSPECIFIED;
  const abs = Math.abs(code);
  if (abs >= 100 && abs <= 109) return ProtoErrorCategory.ERROR_CATEGORY_CONFIGURATION;
  if (abs >= 110 && abs <= 129) return ProtoErrorCategory.ERROR_CATEGORY_MODEL;
  if (abs >= 130 && abs <= 149) return ProtoErrorCategory.ERROR_CATEGORY_INTERNAL;
  if (abs >= 150 && abs <= 179) return ProtoErrorCategory.ERROR_CATEGORY_NETWORK;
  if ((abs >= 180 && abs <= 219) || (abs >= 280 && abs <= 299)) {
    return ProtoErrorCategory.ERROR_CATEGORY_IO;
  }
  if (abs >= 220 && abs <= 229) return ProtoErrorCategory.ERROR_CATEGORY_INTERNAL;
  if (abs >= 230 && abs <= 249) return ProtoErrorCategory.ERROR_CATEGORY_COMPONENT;
  if (abs >= 250 && abs <= 279) return ProtoErrorCategory.ERROR_CATEGORY_VALIDATION;
  if (abs >= 300 && abs <= 319) return ProtoErrorCategory.ERROR_CATEGORY_COMPONENT;
  if (abs >= 320 && abs <= 329) return ProtoErrorCategory.ERROR_CATEGORY_AUTH;
  if (abs >= 330 && abs <= 349) return ProtoErrorCategory.ERROR_CATEGORY_AUTH;
  if (abs >= 350 && abs <= 369) return ProtoErrorCategory.ERROR_CATEGORY_IO;
  if (abs >= 370 && abs <= 379) return ProtoErrorCategory.ERROR_CATEGORY_VALIDATION;
  if (abs >= 380 && abs <= 389) return ProtoErrorCategory.ERROR_CATEGORY_INTERNAL;
  if (abs >= 400 && abs <= 499) return ProtoErrorCategory.ERROR_CATEGORY_COMPONENT;
  if (abs >= 500 && abs <= 599) return ProtoErrorCategory.ERROR_CATEGORY_CONFIGURATION;
  if (abs >= 600 && abs <= 699) return ProtoErrorCategory.ERROR_CATEGORY_COMPONENT;
  if (abs >= 700 && abs <= 799) return ProtoErrorCategory.ERROR_CATEGORY_INTERNAL;
  if (abs >= 800 && abs <= 899) return ProtoErrorCategory.ERROR_CATEGORY_INTERNAL;
  if (abs >= 900 && abs <= 999) return ProtoErrorCategory.ERROR_CATEGORY_INTERNAL;
  return ProtoErrorCategory.ERROR_CATEGORY_UNSPECIFIED;
}

function protoCodeForSDKCode(code: number): ProtoErrorCode {
  const positive = Math.abs(code);
  if (Object.values(ProtoErrorCode).includes(positive)) {
    return positive as ProtoErrorCode;
  }
  return ProtoErrorCode.ERROR_CODE_UNSPECIFIED;
}

function severityForCode(code: number): ProtoErrorSeverity {
  return code === 0
    ? ProtoErrorSeverity.ERROR_SEVERITY_UNSPECIFIED
    : ProtoErrorSeverity.ERROR_SEVERITY_ERROR;
}

function componentForCode(code: number): string {
  if (code === 0) return 'sdk';
  const abs = Math.abs(code);
  if (abs >= 100 && abs <= 109) return 'sdk';
  if (abs >= 110 && abs <= 129) return 'model';
  if (abs >= 130 && abs <= 149) return 'generation';
  if (abs >= 150 && abs <= 179) return 'network';
  if ((abs >= 180 && abs <= 219) || (abs >= 280 && abs <= 299)) return 'storage';
  if (abs >= 220 && abs <= 229) return 'sdk';
  if (abs >= 230 && abs <= 249) return 'component';
  if (abs >= 250 && abs <= 279) return 'validation';
  if (abs >= 300 && abs <= 319) return 'component';
  if (abs >= 320 && abs <= 349) return 'auth';
  if (abs >= 350 && abs <= 369) return 'storage';
  if (abs >= 370 && abs <= 379) return 'validation';
  if (abs >= 380 && abs <= 389) return 'sdk';
  if (abs >= 400 && abs <= 499) return 'component';
  if (abs >= 500 && abs <= 599) return 'sdk';
  if (abs >= 600 && abs <= 699) return 'backend';
  if (abs >= 700 && abs <= 899) return 'sdk';
  if (abs >= 900 && abs <= 999) return 'wasm';
  return 'sdk';
}

export class SDKException extends Error {
  readonly proto: ProtoSDKError;

  constructor(codeOrProto: number | ProtoSDKError, message?: string, details?: string) {
    if (typeof codeOrProto === 'number') {
      const code = codeOrProto;
      const msg = message ?? `SDK error: ${code}`;
      super(msg);
      this.name = 'SDKException';
      this.proto = {
        category: categoryForCode(code),
        code: protoCodeForSDKCode(code),
        cAbiCode: code,
        message: msg,
        nestedMessage: details,
        context: undefined,
        timestampMs: Date.now(),
        severity: severityForCode(code),
        component: componentForCode(code),
        retryable: false,
        remediationHint: '',
        correlationId: '',
      };
    } else {
      super(codeOrProto.message);
      this.name = 'SDKException';
      this.proto = codeOrProto;
    }
  }

  get code(): ProtoErrorCode {
    return this.proto.code;
  }

  get cAbiCode(): number {
    return this.proto.cAbiCode ?? 0;
  }

  get fieldPath(): string | undefined {
    const typed = this.proto.context?.fieldPath;
    return typed && typed.length > 0 ? typed : undefined;
  }

  get recoverySuggestion(): string | undefined {
    switch (this.proto.code) {
      case ProtoErrorCode.ERROR_CODE_NOT_INITIALIZED:
        return 'Initialize the component before using it.';
      case ProtoErrorCode.ERROR_CODE_MODEL_NOT_FOUND:
        return 'Ensure the model is downloaded and the path is correct.';
      case ProtoErrorCode.ERROR_CODE_NETWORK_UNAVAILABLE:
        return 'Check your internet connection and try again.';
      case ProtoErrorCode.ERROR_CODE_INSUFFICIENT_STORAGE:
        return 'Free up storage space and try again.';
      case ProtoErrorCode.ERROR_CODE_INSUFFICIENT_MEMORY:
        return 'Close other applications to free up memory.';
      case ProtoErrorCode.ERROR_CODE_MICROPHONE_PERMISSION_DENIED:
        return 'Grant microphone permission in Settings.';
      case ProtoErrorCode.ERROR_CODE_TIMEOUT:
        return 'Try again or check your connection.';
      case ProtoErrorCode.ERROR_CODE_INVALID_API_KEY:
        return 'Verify your API key is correct.';
      default:
        return undefined;
    }
  }

  log(): void {
    const fields: Record<string, unknown> = {
      error_code: errorCodeToJSON(this.proto.code),
      error_category: errorCategoryToJSON(this.proto.category),
      failure_reason: `[${errorCategoryToJSON(this.proto.category)}] ${errorCodeToJSON(this.proto.code)}`,
    };
    if (this.proto.nestedMessage) fields.underlying_error = this.proto.nestedMessage;
    const sdkFrames = (this.stack ?? '')
      .split('\n')
      .filter((frame) => frame.toLowerCase().includes('runanywhere'))
      .slice(0, 5);
    if (sdkFrames.length > 0) fields.stack_trace = sdkFrames.join('\n');

    const logger = new SDKLogger(errorCategoryToJSON(this.proto.category));
    if (this.proto.code === ProtoErrorCode.ERROR_CODE_CANCELLED) {
      logger.info(this.proto.message, fields);
    } else {
      logger.error(this.proto.message, fields);
    }
  }

  static isSuccess(resultCode: number): boolean {
    return resultCode === 0;
  }

  static fromCode(code: number, message: string, details?: string): SDKException {
    return new SDKException(code, message, details);
  }

  static fromRACResult(resultCode: number, details?: string): SDKException {
    const message = details ? `RACommons error: ${resultCode} (${details})` : `RACommons error: ${resultCode}`;
    return SDKException.fromCode(resultCode, message, details);
  }

  static throwIfError(resultCode: number, details?: string): void {
    if (SDKException.isSuccess(resultCode)) return;
    throw SDKException.fromRACResult(resultCode, details);
  }

  static make(
    code: ProtoErrorCode,
    message: string,
    category: ProtoErrorCategory = ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
    underlying?: Error,
    shouldLog = true,
  ): SDKException {
    const proto: ProtoSDKError = {
      category,
      code,
      cAbiCode: code > 0 && code <= 899 ? -code : 0,
      message,
      nestedMessage: underlying ? String(underlying) : undefined,
      context: undefined,
      timestampMs: Date.now(),
      severity: severityForCode(-code),
      component: componentForCode(-code),
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    const ex = new SDKException(proto);
    if (shouldLog && !isExpected(code)) ex.log();
    return ex;
  }

  static invalidConfiguration(message: string): SDKException {
    return SDKException.make(
      ProtoErrorCode.ERROR_CODE_INVALID_CONFIGURATION,
      message,
      ProtoErrorCategory.ERROR_CATEGORY_CONFIGURATION,
    );
  }

  static timeout(message: string): SDKException {
    return SDKException.make(
      ProtoErrorCode.ERROR_CODE_TIMEOUT,
      message,
      ProtoErrorCategory.ERROR_CATEGORY_NETWORK,
    );
  }

  static networkError(message: string): SDKException {
    return SDKException.make(
      ProtoErrorCode.ERROR_CODE_NETWORK_ERROR,
      message,
      ProtoErrorCategory.ERROR_CATEGORY_NETWORK,
    );
  }

  static notInitialized(message = 'SDK not initialized'): SDKException {
    const proto: ProtoSDKError = {
      category: ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
      code: ProtoErrorCode.ERROR_CODE_NOT_INITIALIZED,
      cAbiCode: -ProtoErrorCode.ERROR_CODE_NOT_INITIALIZED,
      message,
      nestedMessage: undefined,
      context: undefined,
      timestampMs: Date.now(),
      severity: ProtoErrorSeverity.ERROR_SEVERITY_ERROR,
      component: componentForCode(-ProtoErrorCode.ERROR_CODE_NOT_INITIALIZED),
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    return new SDKException(proto);
  }

  static cancelled(message = 'Operation cancelled'): SDKException {
    const proto: ProtoSDKError = {
      category: ProtoErrorCategory.ERROR_CATEGORY_INTERNAL,
      code: ProtoErrorCode.ERROR_CODE_CANCELLED,
      cAbiCode: -380,
      message,
      nestedMessage: undefined,
      context: undefined,
      timestampMs: Date.now(),
      severity: ProtoErrorSeverity.ERROR_SEVERITY_UNSPECIFIED,
      component: 'sdk',
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    return new SDKException(proto);
  }

  static wasmNotLoaded(message = 'WASM module not loaded'): SDKException {
    return SDKException.fromCode(-ProtoErrorCode.ERROR_CODE_WASM_NOT_LOADED, message);
  }

  static modelNotFound(modelId: string): SDKException {
    return SDKException.fromCode(
      -ProtoErrorCode.ERROR_CODE_MODEL_NOT_FOUND,
      `Model not found: ${modelId}`,
    );
  }

  static backendNotAvailable(feature: string, details?: string): SDKException {
    return SDKException.fromCode(
      -ProtoErrorCode.ERROR_CODE_BACKEND_UNAVAILABLE,
      `Backend not available for: ${feature}`,
      details,
    );
  }

  static processingFailed(message: string, details?: string): SDKException {
    const proto: ProtoSDKError = {
      category: ProtoErrorCategory.ERROR_CATEGORY_INTERNAL,
      code: ProtoErrorCode.ERROR_CODE_PROCESSING_FAILED,
      cAbiCode: -ProtoErrorCode.ERROR_CODE_PROCESSING_FAILED,
      message,
      nestedMessage: details,
      context: undefined,
      timestampMs: Date.now(),
      severity: ProtoErrorSeverity.ERROR_SEVERITY_ERROR,
      component: componentForCode(-ProtoErrorCode.ERROR_CODE_PROCESSING_FAILED),
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    return new SDKException(proto);
  }

  static invalidState(message: string, details?: string): SDKException {
    const proto: ProtoSDKError = {
      category: ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
      code: ProtoErrorCode.ERROR_CODE_INVALID_STATE,
      cAbiCode: -ProtoErrorCode.ERROR_CODE_INVALID_STATE,
      message,
      nestedMessage: details,
      context: undefined,
      timestampMs: Date.now(),
      severity: ProtoErrorSeverity.ERROR_SEVERITY_ERROR,
      component: componentForCode(-ProtoErrorCode.ERROR_CODE_INVALID_STATE),
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    return new SDKException(proto);
  }

  static serviceNotAvailable(message: string, details?: string): SDKException {
    const proto: ProtoSDKError = {
      category: ProtoErrorCategory.ERROR_CATEGORY_COMPONENT,
      code: ProtoErrorCode.ERROR_CODE_SERVICE_NOT_AVAILABLE,
      cAbiCode: -ProtoErrorCode.ERROR_CODE_SERVICE_NOT_AVAILABLE,
      message,
      nestedMessage: details,
      context: undefined,
      timestampMs: Date.now(),
      severity: ProtoErrorSeverity.ERROR_SEVERITY_ERROR,
      component: componentForCode(-ProtoErrorCode.ERROR_CODE_SERVICE_NOT_AVAILABLE),
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    return new SDKException(proto);
  }

  static validationFailed(message: string): SDKException;
  static validationFailed(args: { fieldPath: string; message: string; cause?: Error }): SDKException;
  static validationFailed(
    args: string | { fieldPath: string; message: string; cause?: Error },
  ): SDKException {
    if (typeof args === 'string') {
      const proto: ProtoSDKError = {
        category: ProtoErrorCategory.ERROR_CATEGORY_VALIDATION,
        code: ProtoErrorCode.ERROR_CODE_VALIDATION_FAILED,
        cAbiCode: -ProtoErrorCode.ERROR_CODE_VALIDATION_FAILED,
        message: args,
        nestedMessage: undefined,
        context: undefined,
        timestampMs: Date.now(),
        severity: ProtoErrorSeverity.ERROR_SEVERITY_ERROR,
        component: 'validation',
        retryable: false,
        remediationHint: '',
        correlationId: '',
      };
      return new SDKException(proto);
    }
    const proto: ProtoSDKError = {
      category: ProtoErrorCategory.ERROR_CATEGORY_VALIDATION,
      code: ProtoErrorCode.ERROR_CODE_INVALID_ARGUMENT,
      cAbiCode: -259,
      message: args.message,
      nestedMessage: args.cause?.message,
      context: {
        metadata: {},
        sourceFile: undefined,
        sourceLine: undefined,
        operation: undefined,
        fieldPath: args.fieldPath,
      },
      timestampMs: Date.now(),
      severity: ProtoErrorSeverity.ERROR_SEVERITY_ERROR,
      component: 'validation',
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    return new SDKException(proto);
  }

  static fromONNXCode(code: number): SDKException {
    let protoCode: ProtoErrorCode;
    let message: string;
    switch (code) {
      case 0:
        protoCode = ProtoErrorCode.ERROR_CODE_UNKNOWN;
        message = 'Unexpected success code passed to error handler';
        break;
      case -1:
        protoCode = ProtoErrorCode.ERROR_CODE_INITIALIZATION_FAILED;
        message = 'ONNX Runtime initialization failed';
        break;
      case -2:
        protoCode = ProtoErrorCode.ERROR_CODE_MODEL_LOAD_FAILED;
        message = 'Failed to load ONNX model';
        break;
      case -3:
        protoCode = ProtoErrorCode.ERROR_CODE_GENERATION_FAILED;
        message = 'ONNX inference failed';
        break;
      case -4:
        protoCode = ProtoErrorCode.ERROR_CODE_INVALID_STATE;
        message = 'Invalid ONNX handle';
        break;
      case -5:
        protoCode = ProtoErrorCode.ERROR_CODE_INVALID_INPUT;
        message = 'Invalid ONNX parameters';
        break;
      case -6:
        protoCode = ProtoErrorCode.ERROR_CODE_INSUFFICIENT_MEMORY;
        message = 'ONNX Runtime out of memory';
        break;
      case -7:
        protoCode = ProtoErrorCode.ERROR_CODE_NOT_IMPLEMENTED;
        message = 'ONNX feature not implemented';
        break;
      case -8:
        protoCode = ProtoErrorCode.ERROR_CODE_CANCELLED;
        message = 'ONNX operation cancelled';
        break;
      case -9:
        protoCode = ProtoErrorCode.ERROR_CODE_TIMEOUT;
        message = 'ONNX operation timed out';
        break;
      case -10:
        protoCode = ProtoErrorCode.ERROR_CODE_STORAGE_ERROR;
        message = 'ONNX IO error';
        break;
      default:
        protoCode = ProtoErrorCode.ERROR_CODE_UNKNOWN;
        message = `ONNX error code: ${code}`;
        break;
    }
    const proto: ProtoSDKError = {
      category: ProtoErrorCategory.ERROR_CATEGORY_INTERNAL,
      code: protoCode,
      cAbiCode: -protoCode,
      message,
      nestedMessage: undefined,
      context: undefined,
      timestampMs: Date.now(),
      severity: ProtoErrorSeverity.ERROR_SEVERITY_ERROR,
      component: componentForCode(-protoCode),
      retryable: false,
      remediationHint: '',
      correlationId: '',
    };
    return new SDKException(proto);
  }
}

export function isSDKException(error: unknown): error is SDKException {
  return error instanceof SDKException;
}

export function isExpected(code: ProtoErrorCode): boolean {
  return code === ProtoErrorCode.ERROR_CODE_CANCELLED || code === ProtoErrorCode.ERROR_CODE_STREAM_CANCELLED;
}

export type {
  ErrorContext as ProtoErrorContext,
  SDKError as ProtoSDKError,
} from '@runanywhere/proto-ts/errors';
export {
  ErrorCategory as ProtoErrorCategory,
  ErrorCode as ProtoErrorCode,
  ErrorSeverity as ProtoErrorSeverity,
} from '@runanywhere/proto-ts/errors';
